### ISSUE TITLE
Default legend behavior (loc='best') very slow for large amounts of data.

### ISSUE THREAD

**Issue by kbrose** [matplotlib/matplotlib#12120]
### Bug report

**Plotting large amounts of data with legend is slow when legend is using `best` location.**

**Code for reproduction**

```python
import numpy as np
import matplotlib.pyplot as plt
import matplotlib as mpl

# Setup, and create the data to plot
y = np.random.rand(10000000)
# Adapted from https://matplotlib.org/tutorials/introductory/usage.html#performance
mpl.rcParams['path.simplify'] = True
mpl.rcParams['path.simplify_threshold'] = 1.0

plt.plot(y, label='data')
plt.legend(loc='best')
plt.show()

plt.plot(y, label='data')
plt.legend(loc='upper right')
plt.show()


```

**Actual outcome**

The first plot is much slower (`loc='best'`) than the second.

**Matplotlib version**

  * Operating system: Ubuntu 18.04
  * Matplotlib version: 2.1.x and 2.2.x
  * Matplotlib backend (`print(matplotlib.get_backend())`): Qt5Agg, TkAgg
  * Python version: 3.7
  * Jupyter version (if applicable): N/A
  * Other libraries: numpy 1.15.1

Installed via conda.

--------------------------------------------

I think it'd be fine to just note this in the performance section of the User Guide (https://matplotlib.org/tutorials/introductory/usage.html#performance). I could also try looking into speeding up the `best` location finder but I get the feeling that'd be a fairly slow project.

Curious to hear what others think.

**Comment by jklymak** [comment]
I don't think you are going to make it any faster when there are a lot of points.  Its a little too bad `'best'` was chosen as the default because of this.  A documentation PR would be most welcome.  If you can somehow speed it up, that'd be great too, but I think its really hard...

**Comment by kbrose** [comment]
Made a PR (referenced above) with documentation only changes. I know people can be quite opinionated on warnings so I'm going to hold off on that for a while yet. I still have hopes that there may be optimizations to be made.

**Comment by tacaswell** [comment]
Closed by #12203.

Thanks @kbrose !

**Comment by kbrose** [comment]
FYI it looks like there is an unnecessary doubling of runtime. The expensive `_find_best_position` method of `Legend` is called twice during normal operation. Maybe with some memoization things could be sped up -- but I'm guessing it will be tricky to memoize in the context of interactive plotting (hidden state will make it hard to know if the functions output is guaranteed to have stayed the same).

It runs once with a stacktrace like

```
  File "/home/kevin/Documents/github/matplotlib/lib/matplotlib/tests/test_legend.py", line 573, in test_warn_big_data_best_loc
    l.draw(fig.canvas.get_renderer())
  File "/home/kevin/Documents/github/matplotlib/lib/matplotlib/artist.py", line 34, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/home/kevin/Documents/github/matplotlib/lib/matplotlib/legend.py", line 647, in draw
    self._legend_box.draw(renderer)
  File "/home/kevin/Documents/github/matplotlib/lib/matplotlib/offsetbox.py", line 263, in draw
    px, py = self.get_offset(width, height, xdescent, ydescent, renderer)
  File "/home/kevin/Documents/github/matplotlib/lib/matplotlib/offsetbox.py", line 203, in get_offset
    if callable(self._offset)
  File "/home/kevin/Documents/github/matplotlib/lib/matplotlib/legend.py", line 604, in _findoffset
    x, y = self._find_best_position(width, height, renderer)
  File "/home/kevin/Documents/github/matplotlib/lib/matplotlib/legend.py", line 1108, in _find_best_position
```

and again with a stacktrace like
```
  File "/home/kevin/Documents/github/matplotlib/lib/matplotlib/tests/test_legend.py", line 573, in test_warn_big_data_best_loc
    l.draw(fig.canvas.get_renderer())
  File "/home/kevin/Documents/github/matplotlib/lib/matplotlib/artist.py", line 34, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/home/kevin/Documents/github/matplotlib/lib/matplotlib/legend.py", line 635, in draw
    bbox = self._legend_box.get_window_extent(renderer)
  File "/home/kevin/Documents/github/matplotlib/lib/matplotlib/offsetbox.py", line 251, in get_window_extent
    px, py = self.get_offset(w, h, xd, yd, renderer)
  File "/home/kevin/Documents/github/matplotlib/lib/matplotlib/offsetbox.py", line 203, in get_offset
    if callable(self._offset)
  File "/home/kevin/Documents/github/matplotlib/lib/matplotlib/legend.py", line 604, in _findoffset
    x, y = self._find_best_position(width, height, renderer)
  File "/home/kevin/Documents/github/matplotlib/lib/matplotlib/legend.py", line 1108, in _find_best_position
```
